core/src/main/scala/org/http4s/Message.scala

```diff
@@ -414,6 +415,34 @@ sealed abstract case class Request[F[_]](
   override def toString: String =
     s"""Request(method=$method, uri=$uri, headers=${headers.redactSensitive()})"""
 
+  override def hashCode(): Int = MurmurHash3.productHash(this)
+
+  override def equals(that: Any): Boolean = (this eq that.asInstanceOf[Object]) || (that match {
+    case that: Request[_] =>
+      (this.method == that.method) &&
+      (this.uri == that.uri) &&
+      (this.httpVersion == that.httpVersion) &&
+      (this.headers == that.headers) &&
+      (this.body == that.body) &&
+      (this.attributes == that.attributes)
+    case _ => false
+  })
+
+  def canEqual(that: Any) = that.isInstanceOf[Request[F]]
```

Comment:
That `F` is erased and I'm surprised this didn't generate a warning but `_` didn't work.  Maybe `_[_]`?  Ugh.